relay-server/src/actors/store.rs

```diff
@@ -176,6 +135,60 @@ fn make_producer<'a>(
 }
 
 impl StoreForwarder {
+    pub fn start(self) -> Addr<Self> {
+        relay_log::info!("store forwarder started");
+
+        let (tx, mut rx) = mpsc::unbounded_channel::<StoreMessageEnvelope>();
+
+        tokio::spawn(async move {
+            while let Some(message) = rx.recv().await {
+                match message {
+                    StoreMessageEnvelope::StoreEnvelope(msg, responder_tx) => {
+                        let response = self.handle_store_evelope(msg);
+                        responder_tx.send(response).ok();
+                    }
+                }
```

Comment:
Our start/spawn methods will be getting rather large for bigger services like the upstream. Seeing this twice now in the healthcheck and the store actor, I strongly prefer that we move this `match` into a method on the `Envelope` type. 

This allows us to keep the `start` method much smaller and more focused on what it actually does: Orchestrate a run loop and guard resources. How the envelope is being handled is ideally moved to another method. 

You could also add this to a `handle_envelope` method on the service itself, although it will be easier to read and review when it is closer to the envelope's definition as an associated function on the envelope itself.

Reply:
> You could also add this to a `handle_envelope` method on the service itself, although it will be easier to read and review when it is closer to the envelope's definition as an associated function on the envelope itself.

I'd prefer to keep that on the service itself, perhaps a new trait method for the `Service` trait: `fn dispatch_messages(&mut self, message: Self::Messages)` (formerly `Self::Envelope`).  But one reason I wasn't inclined to do this yet is that it now forces you to make the `&mut self` vs `&self` decision globally.

(Aside, to me this is trying to fix future code structure problems that don't yet exist - so there's a lot of (perhaps educated) guessing what the right structure is.  I like postponing things until they are needed)